lms/util/h.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+"""Utility functions for working with the h API."""
+
+from __future__ import unicode_literals
+
+import random
+import string
+
+
+ALLOWED_CHARS = string.ascii_letters + string.digits + "._"
+"""The characters that are allowed in h usernames."""
+
+
+def generate_username(request_params):
+    """Return an h username generated from the given LTI launch request params."""
+
+    full_name = request_params.get("lis_person_name_full") or ""
+    user_id = request_params.get("user_id") or ""
+
+    # Take up to 22 valid characters from full_name.
+    # (h usernames can be up to 30 chars long, and we will be appending 8 chars
+    # from user_id below, leaving us up to 22 chars from full_name).
+    sanitized_full_name = _sanitize_string(full_name, limit=22) or "Anonymous"
+
+    sanitized_user_id = _pad_with_random_characters(
+        _sanitize_string(user_id, limit=8), 8
+    )
+
+    return sanitized_full_name + sanitized_user_id
+
+
+def _sanitize_string(unsanitized_string, limit):
```

Comment:
I'm getting caught up a bit on the naming of this function because it doesn't just sanitize, it truncates. But I don't have a great suggestion, either.

Reply:
I ended up renaming this to:

```python
def n_valid_chars_from(s, n):
        """Return a string of up to ``n`` valid chars from the start of ``s``."""
```

I think it reads well in usage:

```python
s = n_valid_chars_from(s, n=8)
```